### PR TITLE
Transform and inline dsqrt evaluator

### PR DESCRIPTION
**Pull request by simonameng**
Transform to `dsqrt` and inline it when calling `java_lang_StrictMath_sqrt` and `java_lang_Math_sqrt`. Also, anchor and delete the `jclass` child when there are two children in the `dcall java_lang_StrictMath_sqrt`/`dcall java_lang_Math_sqrt` tree.

The changes are for all platforms.

Closes: #5702

### CODE REVIEW

**Comment by simonameng** [comment]
@fjeremic  @dchopra001

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -10041,11 +9932,6 @@ bool J9::X86::TreeEvaluator::VMinlineCallEvaluator(
             }
             break;
 
-         case TR::java_lang_Math_sqrt:
-         case TR::java_lang_StrictMath_sqrt:
```
Comment (fjeremic) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: There is also a similar piece in Power codegen which we need to remove [1].

[1] https://github.com/eclipse/openj9/blob/6887baea4c98f2820f42071c29de5f7a7ce31642/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L12863-L12870

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp [hunk 2]
```diff
@@ -104,6 +104,26 @@ void J9::RecognizedCallTransformer::process_java_lang_StringUTF16_toBytes(TR::Tr
...
+
+      anchorNode(dumpNode, treetop);
+      prepareToReplaceNode(node);
+
+      TR::Node::recreateWithoutProperties(node, TR::dsqrt, 1, valueNode, getSymRefTab()->findOrCreateNewArraySymbolRef(node->getSymbolReference()->getOwningMethodSymbol(comp())));
+      TR::TransformUtil::removeTree(comp(), treetop);
+      }
+
+   else
+      {
```
Comment (fjeremic) [inline runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp]: Extra newline above. Can you explain what this else path is supposed to do? Why are we recreating the call node if the hardware does not support square root?

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp [hunk 3]
```diff
@@ -104,6 +104,26 @@ void J9::RecognizedCallTransformer::process_java_lang_StringUTF16_toBytes(TR::Tr
    treetop->insertAfter(TR::TreeTop::create(comp(), TR::Node::create(node, TR::treetop, 1, newCallNode)));
    }
 
+void J9::RecognizedCallTransformer::process_java_lang_StrictMath_sqrt(TR::TreeTop* treetop, TR::Node* node)
+   {
+   if (!node->getSymbol()->castToResolvedMethodSymbol()->canReplaceWithHWInstr())
```
Comment (fjeremic) [inline runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp]: Not a fan of this API. We should directly test for `TR::Compiler->target.cpu.getSupportsHardwareSQRT()`. Also having the negation here `!` seems like a bug.

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp [hunk 4]
```diff
@@ -421,6 +444,12 @@ void J9::RecognizedCallTransformer::transform(TR::TreeTop* treetop)
       case TR::java_lang_StringUTF16_toBytes:
          process_java_lang_StringUTF16_toBytes(treetop, node);
          break;
+      case TR::java_lang_StrictMath_sqrt:
+         process_java_lang_StrictMath_sqrt(treetop, node);
+         break;
+      case TR::java_lang_Math_sqrt:
+         processIntrinsicFunction(treetop, node, TR::dsqrt);
```
Comment (fjeremic) [inline runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp]: Why do we need to distinguish between these two? Shouldn't they be the same?

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp [hunk 5]
```diff
@@ -367,6 +387,9 @@ bool J9::RecognizedCallTransformer::isInlineable(TR::TreeTop* treetop)
          return !comp()->getOption(TR_DisableMaxMinOptimization);
       case TR::java_lang_StringUTF16_toBytes:
          return !comp()->compileRelocatableCode();
+      case TR::java_lang_StrictMath_sqrt:
+      case TR::java_lang_Math_sqrt:
+         return true;
```
Comment (fjeremic) [inline runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp]: Should we be checking here for `TR::Compiler->target.cpu.getSupportsHardwareSQRT()`? i.e.

```
return TR::Compiler->target.cpu.getSupportsHardwareSQRT();
```

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.hpp [hunk 6]
```diff
@@ -77,6 +77,22 @@ class RecognizedCallTransformer : public OMR::RecognizedCallTransformer
     *     \endcode
     */
    void process_java_lang_StringUTF16_toBytes(TR::TreeTop* treetop, TR::Node* node);
+   /** \brief
+    *     Transforms java/lang/StrictMath.sqrt(D)D into a CodeGen inlined function with equivalent semantics.
```
Comment (fjeremic) [inline runtime/compiler/optimizer/J9RecognizedCallTransformer.hpp]: Is there a reason we cannot handle `java/lang/[Strict]Math.sqrt(F)F` in a similar way?

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp (hunk 4 above)
Comment (simonameng) [reply runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp]: The reason we distinguish them is that `TR::java_lang_Math_sqrt` is not a native function. There won't be a `jclass` child with it. And `TR::java_lang_StrictMath_sqrt` is a native function with two children. So I want to process the two-children case separately.

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.hpp (hunk 6 above)
Comment (dchopra001) [reply runtime/compiler/optimizer/J9RecognizedCallTransformer.hpp]: Is there such an overload? I don't see one in the java docs. The only one I see is for `double` type.

Comment (fjeremic) [reply runtime/compiler/optimizer/J9RecognizedCallTransformer.hpp]: You're right! TIL.

**Comment by simonameng** [comment]
@fjeremic  Should be good to go now.

**Comment by fjeremic** [comment]
Jenkins test extended all jdk8

**Pull request merged**